grpc/read_job.cpp

```diff
@@ -0,0 +1,167 @@
+	const elliptics::data_pointer &data, size_t &offset) {
+
+	if (offset == data.size() || builder.GetSize() >= GRPC_MAX_MESSAGE_SIZE) {
+		return 0;
+	}
+
+	size_t max_part_size = GRPC_MAX_MESSAGE_SIZE - builder.GetSize();
+	size_t size_left = data.size() - offset;
+	size_t part_size = std::min(size_left, max_part_size);
+
+	auto part_ptr = static_cast<const uint8_t *>(data.skip(offset).data());
+	auto res = builder.CreateVector(part_ptr, part_size);
+	offset += part_size;
+	return res;
+}
+
+read_job_t::rpc_response_t serialize_part(const read_job_t::response_t &response, bool first,
+	std::size_t &json_offset, std::size_t &data_offset) {
+
+	flatbuffers::grpc::MessageBuilder builder;
+
+	auto header = first
+	              ? serialize_header(builder, response)
+	              : flatbuffers::Offset<fb_grpc_dnet::ReadResponseHeader>();
+
+	builder.Finish(fb_grpc_dnet::CreateReadResponse(
+		builder,
+		header,
+		put_data_part(builder, response.json, json_offset),
+		put_data_part(builder, response.data.in_memory, data_offset)
+	));
+	return builder.ReleaseMessage<fb_grpc_dnet::ReadResponse>();
+}
+
+} // namespace
+
+class read_job_t::responder_t {
+public:
+	explicit responder_t(read_job_t &parent)
+	: parent_(parent) {
+	}
+
+	void reply(std::unique_ptr<response_t> response) {
+		parent_.send_response(std::move(response));
+	}
+
+private:
+	read_job_t &parent_;
+};
+
+read_job_t::read_job_t(dnet_node &node, ::grpc::ServerCompletionQueue &completion_queue,
+	fb_grpc_dnet::Elliptics::AsyncService &service)
+: async_writer_(&ctx_)
+, node_(node)
+, completion_queue_(completion_queue)
+, async_service_(service) {
+    proceed(true);
+}
+
+void read_job_t::proceed(bool ok) {
+	// TODO: catch exceptions here
+	if (!ok) {
+	    // TODO log
+	    delete this;
+	    return;
+	}
+	switch (state_) {
+	case state::CREATE:
+		state_ = state::REQUEST_RECEIVED;
+		async_service_.RequestRead(&ctx_, &rpc_request_, &async_writer_, &completion_queue_,
+		                           &completion_queue_, this);
+		break;
+	case state::REQUEST_RECEIVED:
+		new read_job_t(node_, completion_queue_, async_service_);
+		push_request();
+		break;
+	case state::RESPONSE_PARTIAL_COMPLETE:
+		send_next(false /*first*/);
+		break;
+	case state::FINISH:
+		delete this;
+		break;
+	}
+}
+
+void read_job_t::push_request() {
+	// TODO: Use constructor or clear memory for request
+	auto request = new request_t;
+	deserialize(rpc_request_, *request);
+	request->deadline = to_dnet_time(ctx_.deadline());
+
+	// TODO add to req
+	// auto responder = new responder_t(*this);
+
+	auto req = new dnet_io_req;
+	memset(req, 0, sizeof(dnet_io_req));
+	req->st = node_.st;
+	req->header = static_cast<void*>(&request->cmd);
+	req->n2_msg = request;
+
+	dnet_schedule_io(&node_, req);
+}
+
+void read_job_t::send_response(std::unique_ptr<response_t> response) {
+	response_ = std::move(response);
+	state_ = state::RESPONSE_PARTIAL_COMPLETE;
```

Comment:
Sorry, I forget about one thing. We had discussion about `state::CREATE` and all state machine in one place. Here is a part of state machine, that is outside.

In my result opinion, `state` is required to choose action to handle completion event. `state::CREATE` doesn't relate to any completion event, so it is outsider. I think it was mistake to introduce this state. I google examples response was sent directly inline on completion event, and only due to this coincidence all the actions was situated in `switch`. It's not good pattern to pass all the actions through single bottle neck, and the split them again. If you're not against, I could rework this by myself, in some next PR.

Reply:
Ok, I'm not against.